File: 구명보트(재귀).java
import java.util.*;
import java.util.stream.*;
class Solution {
    private int boat;
    private int limit;
    private void method(List<Integer> list){
        boolean finish = false;
        if(list.isEmpty()){
            
            return;
        }
        for(int i = 1 ; i < list.size(); i++){
            
            if(list.get(0) + list.get(i) <= this.limit){
                this.boat += 1;
                list.remove(i);
                list.remove(0);
                finish = true;
            
                break;
            }   
        }
        if(!finish){
            list.remove(0);
            this.boat += 1;
        }
        method(list);
    }
    public int solution(int[] people, int limit) {
        this.limit = limit;
        
        List<Integer> sorted = Arrays.stream(people)
            .boxed()
            .sorted(Collections.reverseOrder())
            .collect(Collectors.toList());
        method(sorted);
        
        return this.boat;
    }
}

File: forJS/구명 보트(재귀).js
function solution(people, limit) {
  let boat = 0;

  function method(list) {
    let finish = false;
    if (list.length === 0) {
      return;
    }
    for (let i = 1; i < list.length; i++) {
      if (list[0] + list[i] <= limit) {
        boat += 1;
        list.splice(i, 1);
        list.shift();
        finish = true;
        break;
      }
    }
    if (!finish) {
      list.shift();
      boat += 1;
    }
    method(list);
  }

  const sorted = people.slice().sort((a, b) => b - a);
  method(sorted);

  return boat;
}
